consensus/signatures.go

```diff
@@ -0,0 +1,198 @@
+// seg faults by throwing covered fields objects at the state which point to
+// nonexistent objects in the transaction.
+
+// Each input has a list of public keys and a required number of signatures.
+// InputSignatures keeps track of which public keys have been used and how many
+// more signatures are needed.
+type InputSignatures struct {
+	RemainingSignatures uint64
+	PossibleKeys        []SiaPublicKey
+	UsedKeys            map[uint64]struct{}
+	Index               int
+}
+
+// sortedUnique checks that all of the elements in a []unit64 are sorted and
+// without repeates, and also checks that the largest element is less than or
+// equal to the biggest allowed element.
+func sortedUnique(elems []uint64, biggestAllowed int) (err error) {
+	biggest := elems[0]
+	for _, elem := range elems[1:] {
+		if elem <= biggest {
+			err = errors.New("covered fields sorting violation")
+			return
+		}
+	}
+	if int(biggest) > biggestAllowed {
+		err = errors.New("covered fields indexing violation")
+		return
+	}
+	return
+}
+
+// validCoveredFields makes sure that all covered fields objects in the
+// signatures follow the rules. This means that if `WholeTransaction` is set to
+// true, all fields except for `Signatures` must be empty. All fields must be
+// sorted numerically, and there can be no repeats.
+func (t Transaction) validCoveredFields() (err error) {
+	for _, sig := range t.Signatures {
+		// Check that all fields are empty if `WholeTransaction` is set.
+		cf := sig.CoveredFields
+		if cf.WholeTransaction {
+			if len(cf.Inputs) != 0 ||
+				len(cf.MinerFees) != 0 ||
+				len(cf.Outputs) != 0 ||
+				len(cf.FileContracts) != 0 ||
+				len(cf.StorageProofs) != 0 ||
+				len(cf.SiafundInputs) != 0 ||
+				len(cf.SiafundOutputs) != 0 ||
+				len(cf.ArbitraryData) != 0 {
+				err = errors.New("whole transaction is set but not all fields besides signatures are empty")
+				return
+			}
+		}
+
+		// Check that all fields are sorted, and without repeat values, and
+		// that all elements point to objects that exists within the
+		// transaction.
+		err = sortedUnique(cf.Inputs, len(cf.Inputs)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.MinerFees, len(cf.MinerFees)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.Outputs, len(cf.Outputs)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.FileContracts, len(cf.FileContracts)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.StorageProofs, len(cf.StorageProofs)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.SiafundInputs, len(cf.SiafundInputs)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.SiafundOutputs, len(cf.SiafundOutputs)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.ArbitraryData, len(cf.ArbitraryData)-1)
+		if err != nil {
+			return
+		}
+		err = sortedUnique(cf.Signatures, len(cf.Signatures)-1)
+		if err != nil {
+			return
+		}
```

Comment:
just that the `if` statement doesn't check signatures, so you couldn't do the `for` loop thing. (Unless you resliced it, but ugh. Just leave it.)